### PR TITLE
REGEXP_SUBSTR

### PR DESCRIPTION
**Pull request by Technay**

### CODE REVIEW

File: core/src/main/java/org/apache/calcite/sql/fun/SqlLibraryOperators.java [hunk 1]
```diff
@@ -124,7 +124,7 @@ private SqlLibraryOperators() {
    *
    * <p>It has similar semantics to standard SQL's
    * {@link SqlStdOperatorTable#SUBSTRING} function but different syntax. */
-  @LibraryOperator(libraries = {ORACLE})
+  @LibraryOperator(libraries = {ORACLE, BIGQUERY})
```
Comment (stutigupta028) [inline core/src/main/java/org/apache/calcite/sql/fun/SqlLibraryOperators.java]: Add Hive And Spark as well

File: core/src/main/java/org/apache/calcite/sql/dialect/BigQuerySqlDialect.java [hunk 2]
```diff
@@ -220,11 +222,69 @@ public BigQuerySqlDialect(SqlDialect.Context context) {
...
+        call.operand(1).unparse(writer, leftPrec, rightPrec);
+      }
+      writer.endFunCall(regexFrame);
+      unparseOccurrenceNode(writer, call);
+      break;
+    }
+  }
+
+  private void unparseOccurrenceNode(SqlWriter writer, SqlCall call) {
+    String occNode = call.operand(3).toString();
```
Comment (stutigupta028) [inline core/src/main/java/org/apache/calcite/sql/dialect/BigQuerySqlDialect.java]: Give a better variable name... occNode is not meaningful

File: core/src/main/java/org/apache/calcite/sql/dialect/BigQuerySqlDialect.java [hunk 3]
```diff
@@ -220,11 +222,69 @@ public BigQuerySqlDialect(SqlDialect.Context context) {
...
+      writer.endFunCall(regexFrame);
+      unparseOccurrenceNode(writer, call);
+      break;
+    case 5:
+      regexFrame = writer.startFunCall("REGEXP_EXTRACT_ALL");
+      unparseSubstr(writer, call, leftPrec, rightPrec);
+      writer.sep(",");
+      if (call.operand(4).toString().equals("'i'")) {
+        writer.literal("'(?i)".concat(call.operand(1).toString()));
+      } else {
```
Comment (stutigupta028) [inline core/src/main/java/org/apache/calcite/sql/dialect/BigQuerySqlDialect.java]: there are I guess 5 options and not just 2, so every other case will not go to else. Please check the Teradata document once

File: core/src/main/java/org/apache/calcite/sql/dialect/BigQuerySqlDialect.java [hunk 4]
```diff
@@ -220,11 +222,69 @@ public BigQuerySqlDialect(SqlDialect.Context context) {
...
+      break;
+    }
+  }
+
+  private void unparseOccurrenceNode(SqlWriter writer, SqlCall call) {
+    String occNode = call.operand(3).toString();
+    writer.literal("[OFFSET(" + occNode + " - 1)]");
+  }
+
+  private void unparseSubstr(SqlWriter writer, SqlCall call, int leftPrec, int rightPrec) {
```
Comment (stutigupta028) [inline core/src/main/java/org/apache/calcite/sql/dialect/BigQuerySqlDialect.java]: why was there a need to unparse SUBSTR here again?

File: core/src/main/java/org/apache/calcite/sql/dialect/BigQuerySqlDialect.java [hunk 5]
```diff
@@ -220,11 +222,69 @@ public BigQuerySqlDialect(SqlDialect.Context context) {
...
     }
   }
 
+  private void unparseRegexSubstr(SqlWriter writer, SqlCall call, int leftPrec, int rightPrec) {
+    final SqlWriter.Frame regexFrame;
+    switch (call.operandCount()) {
+    case 2:
+      regexFrame = writer.startFunCall("REGEXP_EXTRACT");
+      for (SqlNode operand : call.getOperandList()) {
+        writer.sep(",");
```
Comment (stutigupta028) [inline core/src/main/java/org/apache/calcite/sql/dialect/BigQuerySqlDialect.java]: Can we do it in a better way , like 
  SqlSyntax.FUNCTION.unparse(writer, REGEXP_EXTRACT, call, leftPrec, rightPrec);

File: core/src/main/java/org/apache/calcite/sql/dialect/BigQuerySqlDialect.java [hunk 6]
```diff
@@ -220,11 +222,69 @@ public BigQuerySqlDialect(SqlDialect.Context context) {
...
+        operand.unparse(writer, leftPrec, rightPrec);
+      }
+      writer.endFunCall(regexFrame);
+      break;
+    case 3:
+      regexFrame = writer.startFunCall("REGEXP_EXTRACT");
```
Comment (stutigupta028) [inline core/src/main/java/org/apache/calcite/sql/dialect/BigQuerySqlDialect.java]: I guess we can optimize every case other in a similar way. Except for case 5th where you are modifying the operand

File: core/src/main/java/org/apache/calcite/sql/fun/SqlLibraryOperators.java (hunk 1 above)
Comment (Technay) [reply core/src/main/java/org/apache/calcite/sql/fun/SqlLibraryOperators.java]: Once bigquery is handled with all variations, I will add spark and hive in as well.

File: core/src/main/java/org/apache/calcite/sql/dialect/BigQuerySqlDialect.java [hunk 7]
```diff
@@ -220,11 +222,69 @@ public BigQuerySqlDialect(SqlDialect.Context context) {
...
       break;
+    case REGEXP_SUBSTR:
+      unparseRegexSubstr(writer, call, leftPrec, rightPrec);
+      break;
     default:
       super.unparseCall(writer, call, leftPrec, rightPrec);
     }
   }
 
+  private void unparseRegexSubstr(SqlWriter writer, SqlCall call, int leftPrec, int rightPrec) {
```
[the remark this answers is not in the document]
Comment (Technay) [reply core/src/main/java/org/apache/calcite/sql/dialect/BigQuerySqlDialect.java]: Sure

File: core/src/main/java/org/apache/calcite/sql/dialect/BigQuerySqlDialect.java (hunk 2 above)
Comment (Technay) [reply core/src/main/java/org/apache/calcite/sql/dialect/BigQuerySqlDialect.java]: occurrenceNode?

File: core/src/main/java/org/apache/calcite/sql/dialect/BigQuerySqlDialect.java (hunk 3 above)
Comment (Technay) [reply core/src/main/java/org/apache/calcite/sql/dialect/BigQuerySqlDialect.java]: I am aware about that. There are total 6 valid arguments and some arguments can be combined increasing unique valid arguments. We need to discuss if we are going to handle all of them.

File: core/src/main/java/org/apache/calcite/sql/dialect/BigQuerySqlDialect.java (hunk 4 above)
Comment (Technay) [reply core/src/main/java/org/apache/calcite/sql/dialect/BigQuerySqlDialect.java]: I have just separated out common code into a method

File: core/src/main/java/org/apache/calcite/sql/dialect/BigQuerySqlDialect.java (hunk 5 above)
Comment (Technay) [reply core/src/main/java/org/apache/calcite/sql/dialect/BigQuerySqlDialect.java]: I am unaware of this class. Will have a look at it.

File: core/src/main/java/org/apache/calcite/sql/fun/SqlJsonRemoveFunction.java [hunk 8]
```diff
@@ -34,7 +34,7 @@
  */
 public class SqlJsonRemoveFunction extends SqlFunction {
 
-  public SqlJsonRemoveFunction() {
+  public    SqlJsonRemoveFunction() {
```
Comment (stutigupta028) [inline core/src/main/java/org/apache/calcite/sql/fun/SqlJsonRemoveFunction.java]: Remove extra spaces

File: core/src/main/java/org/apache/calcite/sql/fun/SqlLibraryOperators.java [hunk 9]
```diff
@@ -182,6 +187,81 @@ private SqlLibraryOperators() {
   @LibraryOperator(libraries = {MYSQL, ORACLE})
   public static final SqlFunction REGEXP_REPLACE = new SqlRegexpReplaceFunction();
 
+  @LibraryOperator(libraries = BIGQUERY)
+  public static final SqlFunction REGEXP_SUBSTR = new SqlRegexpSubstrFunction();
+
+  /**
+   * The REGEXP_EXTRACT(source_string, regex_pattern) returns the first substring in source_string
```
Comment (stutigupta028) [inline core/src/main/java/org/apache/calcite/sql/fun/SqlLibraryOperators.java]: Move the comments of REGEXP_EXTRACT and REGEXP_EXTRACT_ALL above their function definition

File: core/src/main/java/org/apache/calcite/sql/fun/SqlLibraryOperators.java [hunk 10]
```diff
@@ -182,6 +187,81 @@ private SqlLibraryOperators() {
...
+  SqlTypeTransforms.TO_NULLABLE),
+    null, OperandTypes.REGEX_REPLACE_OPERAND_TYPE,
+  SqlFunctionCategory.STRING);
+
+  @LibraryOperator(libraries = {BIGQUERY, HIVE, SPARK})
+  public static final SqlFunction REGEXP_EXTRACT_ALL = new SqlFunction("REGEXP_EXTRACT_ALL",
+    SqlKind.OTHER_FUNCTION,
+    ReturnTypes.cascade(ReturnTypes.explicit(SqlTypeName.VARCHAR),
+      SqlTypeTransforms.TO_NULLABLE),
+    null, OperandTypes.REGEX_REPLACE_OPERAND_TYPE,
```
Comment (stutigupta028) [inline core/src/main/java/org/apache/calcite/sql/fun/SqlLibraryOperators.java]: Change the operand type for REGEXP_EXTRACT_ALL and REGEXP_EXTRACT

File: core/src/main/java/org/apache/calcite/sql/fun/SqlRegexpReplaceFunction.java [hunk 11]
```diff
@@ -39,8 +39,7 @@ public SqlRegexpReplaceFunction() {
         SqlKind.OTHER_FUNCTION,
         ReturnTypes.cascade(ReturnTypes.explicit(SqlTypeName.VARCHAR),
             SqlTypeTransforms.TO_NULLABLE),
-        null,
-        null,
+        null, OperandTypes.REGEX_REPLACE_OPERAND_TYPE,
```
Comment (stutigupta028) [inline core/src/main/java/org/apache/calcite/sql/fun/SqlRegexpReplaceFunction.java]: revert this change

File: core/src/main/java/org/apache/calcite/sql/fun/SqlStdOperatorTable.java [hunk 12]
```diff
@@ -1761,6 +1761,10 @@ public SqlSyntax getSyntax() {
 
   public static final SqlFunction NULLIF = new SqlNullifFunction();
 
+  public static final SqlFunction REGEXP_REPLACE = SqlLibraryOperators.REGEXP_REPLACE;
```
Comment (stutigupta028) [inline core/src/main/java/org/apache/calcite/sql/fun/SqlStdOperatorTable.java]: remove REGEXP_REPLACE from Standard from this request

File: core/src/test/java/org/apache/calcite/rel/rel2sql/RelToSqlConverterTest.java [hunk 13]
```diff
@@ -114,7 +114,7 @@ private static Planner getPlanner(List<RelTraitDef> traitDefs,
       SqlToRelConverter.Config sqlToRelConf, Program... programs) {
     final SchemaPlus rootSchema = Frameworks.createRootSchema(true);
     final FrameworkConfig config = Frameworks.newConfigBuilder()
-        .parserConfig(parserConfig)
+        .parserConfig( parserConfig)
```
Comment (stutigupta028) [inline core/src/test/java/org/apache/calcite/rel/rel2sql/RelToSqlConverterTest.java]: remove extra space
